### DISCUSSION TITLE
MPC trajectory tracking controller for a bicycle model

### DISCUSSION THREAD

**Discussion by josyulakrishna** [python-control/python-control#841] [Q&A]
Hi, 
I have so far followed the examples in the documentation, and I am trying to implement a simple state-space MPC-based trajectory tracking controller based on the kinematic bicycle model shown in figure.1 

![image](https://user-images.githubusercontent.com/2005601/210692921-9f87fb87-6251-4efc-a071-be233b9f3a79.png)
Fig.1 Model
I have a reference trajectory for the states (x_ref(t),y_ref(t), v_ref(t), psi_ref(t)) obtained from a cubic polynomial and would want to optimize a similar objective given in Fig.2 
![image](https://user-images.githubusercontent.com/2005601/210693246-8f21ee85-0be7-484e-8720-8af3b966aec9.png)
where 
![image](https://user-images.githubusercontent.com/2005601/210693291-2c0a38a0-ee7b-4298-a6d4-be72ff3e7791.png)

I have understood that I can encode the bicycle model as `NonlinearIOSystem` as 

```

 def vehicle_update(t,x,u, params):
    # x = [x,y,v,yaw]
    v = x[2]
    psi = x[3]
    a = u[0]
    delta = np.clip(u[1],-np.deg2rad(45), np.deg2rad(45))
    xdot =  v * np.cos(psi)
    ydot =  v * np.sin(psi)
    psidot = (v /P.WB) * np.tan(delta)
    vdot = np.clip( a * P.dt, P.speed_min, P.speed_min)
    return np.array([xdot, ydot, psidot, vdot])

def vehicle_output(t,x,u, params):
    return x

def solve_nonlinear_mpc(z_ref, x):
    vehicle = ct.NonlinearIOSystem(
        vehicle_update, vehicle_output, states=4, name='vehicle',
        inputs=['a', 'delta'], outputs=['x', 'y', 'v', 'psi'], dt=0.2)
   #what goes next?##
```

I now have to code the cost function which essentially computes `sum_T_steps((x(t)-x_ref(t))^2 +(y(t)-y_ref(t))^2 +(v(t)-v_ref(t))^2+((psi(t)-psi_ref(t))^2)` I'm really not understanding how to proceed from here how do I pass the reference trajectory to `opt.quadratic_cost` function? or do I have to pass a cost function handle, if so, how do I handle the computation of `x(t)` using the `vehicle_update` function in the cost function? 

 Thank you.

**Comment by murrayrm** [comment]
I don't think you can do this in the current version of the python-control toolbox.  The problem is that the current setup does not allow a time-varying integral cost.  So there is no way to implement the $\eta_{k,t} - \eta_{\text{ref}_{k,t}}$ expression, at least that I can think of.  The $\Delta u_{k,t}$ term is also problematic, since the integral cost as written can only depend on the current state and input (though there might be a way to fudge this but using costs that are negative?).

**Reply by josyulakrishna** [reply]
I see thank you for the clarification, Prof. Murray.
